src/SpacetimeDBClient.cs

```diff
@@ -14,9 +14,64 @@
 
 namespace SpacetimeDB
 {
-    public abstract class SpacetimeDBClientBase<ReducerEvent>
-        where ReducerEvent : ReducerEventBase
+    public sealed class DbConnectionBuilder<DbConnection, EventContext>
+        where DbConnection : DbConnectionBase<DbConnection, EventContext>, new()
+        where EventContext : class, IEventContext
     {
+        readonly DbConnection conn = new();
+
+        string? uri;
+        string? nameOrAddress;
+        string? token;
+
+        public DbConnection Build() {
+            if (uri == null) {
+                throw new InvalidOperationException("Building DbConnection with a null uri. Call WithUri() first.");
+            }
+            if (nameOrAddress == null) {
+                throw new InvalidOperationException("Building DbConnection with a null nameOrAddress. Call WithModuleName() first.");
+            }
+            conn.Connect(token, uri, nameOrAddress);
+            return conn;
+        }
+
+        public DbConnectionBuilder<DbConnection, EventContext> WithUri(string uri) {
+            this.uri = uri;
+            return this;
+        }
+
+        public DbConnectionBuilder<DbConnection, EventContext> WithModuleName(string nameOrAddress) {
+            this.nameOrAddress = nameOrAddress;
+            return this;
+        }
+
+        public DbConnectionBuilder<DbConnection, EventContext> WithCredentials(in (Identity identity, string token)? creds) {
+            token = creds?.token;
+            return this;
+        }
+
+        public DbConnectionBuilder<DbConnection, EventContext> OnConnect(Action<Identity, string> cb) {
+            conn.onConnect += cb;
+            return this;
+        }
+
+        public DbConnectionBuilder<DbConnection, EventContext> OnConnectError(Action<WebSocketError?, string> cb) {
+            conn.webSocket.OnConnectError += (a, b) => cb.Invoke(a, b);
```

Comment:
Is this not the same as

```suggestion
            conn.webSocket.OnConnectError += cb;
```

Reply:
No, it cannot convert the Action<> into the Handler delegate even if the arguments are the same.

Reply:
I guess you could accept the handler type here?

Reply:
Not important anyway.